File: backend/services/pricing.py
from decimal import Decimal, InvalidOperation, ROUND_HALF_UP

MONEY_QUANT = Decimal("0.01")
RATIO_QUANT = Decimal("0.0001")
DEFAULT_MARKUP = Decimal("2.00")
DEFAULT_PACKAGING_COST = Decimal("9.34")

PAYMENT_FEES = {
    "preco_pix": Decimal("0.00"),
    "preco_debito": Decimal("1.37"),
    "preco_credito_vista": Decimal("3.15"),
    "preco_credito_2x": Decimal("5.39"),
    "preco_credito_3x": Decimal("6.12"),
    "preco_credito_4x": Decimal("6.85"),
    "preco_credito_5x": Decimal("7.57"),
    "preco_credito_6x": Decimal("8.28"),
    "preco_credito_7x": Decimal("8.99"),
    "preco_credito_8x": Decimal("9.69"),
    "preco_credito_9x": Decimal("10.38"),
    "preco_credito_10x": Decimal("11.06"),
    "preco_credito_11x": Decimal("11.74"),
    "preco_credito_12x": Decimal("12.40"),
}

CALCULATED_PRICE_FIELDS = tuple(PAYMENT_FEES.keys()) + (
    "custo_total",
    "lucro_pix",
    "margem_pix",
)


def money(value, *, field="valor", required=True, default=None, minimum=Decimal("0.00")):
    if value in (None, ""):
        if default is not None:
            value = default
        elif required:
            raise ValueError(f"Campo obrigatorio: {field}")
        else:
            return None
    try:
        number = Decimal(str(value)).quantize(MONEY_QUANT, rounding=ROUND_HALF_UP)
    except (InvalidOperation, TypeError, ValueError) as exc:
        raise ValueError(f"{field} invalido") from exc
    if number < Decimal(str(minimum)):
        raise ValueError(f"{field} nao pode ser negativo")
    return number


def ratio(value, *, field="markup", default=DEFAULT_MARKUP):
    if value in (None, ""):
        value = default
    try:
        number = Decimal(str(value)).quantize(Decimal("0.01"), rounding=ROUND_HALF_UP)
    except (InvalidOperation, TypeError, ValueError) as exc:
        raise ValueError(f"{field} invalido") from exc
    if number <= 0:
        raise ValueError(f"{field} deve ser maior que zero")
    return number


def calculate_pricing(custo_peca, custo_embalagem=None, markup=None):
    cost_piece = money(custo_peca, field="custo_peca")
    packaging = money(
        custo_embalagem,
        field="custo_embalagem",
        default=DEFAULT_PACKAGING_COST,
    )
    markup_value = ratio(markup)
    total_cost = (cost_piece + packaging).quantize(MONEY_QUANT, rounding=ROUND_HALF_UP)

    result = {
        "custo_peca": cost_piece,
        "custo_embalagem": packaging,
        "custo_total": total_cost,
        "markup": markup_value,
    }
    for field, fee in PAYMENT_FEES.items():
        denominator = Decimal("1.00") - (fee / Decimal("100"))
        result[field] = (total_cost * markup_value / denominator).quantize(
            MONEY_QUANT,
            rounding=ROUND_HALF_UP,
        )

    pix_price = result["preco_pix"]
    pix_fee = PAYMENT_FEES["preco_pix"]
    profit = (pix_price - total_cost - (pix_price * pix_fee / Decimal("100"))).quantize(
        MONEY_QUANT,
        rounding=ROUND_HALF_UP,
    )
    result["lucro_pix"] = profit
    result["margem_pix"] = (
        (profit / pix_price).quantize(RATIO_QUANT, rounding=ROUND_HALF_UP)
        if pix_price
        else Decimal("0.0000")
    )
    return result


def apply_pricing(product, *, custo_peca, custo_embalagem=None, markup=None):
    values = calculate_pricing(custo_peca, custo_embalagem, markup)
    for field, value in values.items():
        setattr(product, field, value)
    product.price = values["preco_pix"]
    return values

File: frontend/js/vj-admin/pricing.js
(function () {
    const fees = {
        preco_pix: 0,
        preco_debito: 1.37,
        preco_credito_vista: 3.15,
        preco_credito_2x: 5.39,
        preco_credito_3x: 6.12,
        preco_credito_4x: 6.85,
        preco_credito_5x: 7.57,
        preco_credito_6x: 8.28,
        preco_credito_7x: 8.99,
        preco_credito_8x: 9.69,
        preco_credito_9x: 10.38,
        preco_credito_10x: 11.06,
        preco_credito_11x: 11.74,
        preco_credito_12x: 12.40,
    };

    const labels = {
        preco_pix: 'Pix',
        preco_debito: 'Debito',
        preco_credito_vista: 'Credito 1x',
        preco_credito_2x: 'Credito 2x',
        preco_credito_3x: 'Credito 3x',
        preco_credito_4x: 'Credito 4x',
        preco_credito_5x: 'Credito 5x',
        preco_credito_6x: 'Credito 6x',
        preco_credito_7x: 'Credito 7x',
        preco_credito_8x: 'Credito 8x',
        preco_credito_9x: 'Credito 9x',
        preco_credito_10x: 'Credito 10x',
        preco_credito_11x: 'Credito 11x',
        preco_credito_12x: 'Credito 12x',
    };

    function roundMoney(value) {
        return Math.round((Number(value) + Number.EPSILON) * 100) / 100;
    }

    function calculate(custoPeca, custoEmbalagem = 9.34, markup = 2) {
        const piece = Number(custoPeca || 0);
        const packaging = Number(custoEmbalagem || 0);
        const markupValue = Number(markup || 2);
        const custoTotal = roundMoney(piece + packaging);
        const result = {
            custo_peca: roundMoney(piece),
            custo_embalagem: roundMoney(packaging),
            custo_total: custoTotal,
            markup: markupValue,
        };
        Object.entries(fees).forEach(([field, fee]) => {
            result[field] = roundMoney((custoTotal * markupValue) / (1 - fee / 100));
        });
        result.lucro_pix = roundMoney(result.preco_pix - custoTotal);
        result.margem_pix = result.preco_pix ? result.lucro_pix / result.preco_pix : 0;
        return result;
    }

    window.VJAdminPricing = { fees, labels, calculate };
})();
